src/envs/behavior_options.py

```diff
@@ -0,0 +1,976 @@
+            x,
+            y,
+            z,
+            euler_angles[0],
+            euler_angles[1],
+            euler_angles[2],
+        ],
+        hand_limits=((minx, miny, minz), (maxx, maxy, maxz)),
+        obstacles=get_body_ids(env,
+                               include_self=True,
+                               grasping_with_right=True),
+        rng=rng,
+    )
+    p.restoreState(state)
+
+    # NOTE: This below line is *VERY* important after the
+    # pybullet state is restored. The hands keep an internal
+    # track of their state, and if we don't reset this
+    # state to mirror the actual pybullet state, the hand will
+    # think it's elsewhere and update incorrectly accordingly
+    env.robots[0].parts["right_hand"].set_position(
+        env.robots[0].parts["right_hand"].get_position())
+    env.robots[0].parts["left_hand"].set_position(
+        env.robots[0].parts["left_hand"].get_position())
+
+    # If RRT planning fails, fail and return None
+    if plan is None:
+        print(f"PRIMITIVE: grasp {obj.name} fail, failed" +
+              " to find plan to continuous params" + f" {grasp_offset}")
+        return None
+
+    # Grasping Phase 2: Move along the vector from the
+    # position the hand ends up in to the object and
+    # then try to grasp.
+    hand_pos = plan[-1][0:3]
+    hand_orn = plan[-1][3:6]
+    # Get the closest point on the object's bounding
+    # box at which we can try to put the hand
+    closest_point_on_aabb = get_closest_point_on_aabb(hand_pos, lo, hi)
+    delta_pos_to_obj = [
+        closest_point_on_aabb[0] - hand_pos[0],
+        closest_point_on_aabb[1] - hand_pos[1],
+        closest_point_on_aabb[2] - hand_pos[2],
+    ]
+    # we want to accomplish the motion in 25 timesteps
+    # NOTE: this is an arbitrary choice
+    delta_step_to_obj = [delta_pos / 25.0 for delta_pos in delta_pos_to_obj]
+
+    # move the hand along the vector to the object until it
+    # touches the object
+    for _ in range(25):
+        new_hand_pos = [
+            hand_pos[0] + delta_step_to_obj[0],
+            hand_pos[1] + delta_step_to_obj[1],
+            hand_pos[2] + delta_step_to_obj[2],
+        ]
+        plan.append(new_hand_pos + list(hand_orn))
+        hand_pos = new_hand_pos
+
+    # Setup two booleans to be used as 'memory', as well as
+    # a 'reversed' plan to be used
+    # for our option that's defined below
+    reversed_plan = list(reversed(plan[:]))
+    plan_executed_forwards = False
+    tried_closing_gripper = False
+
+    def graspObjectOption(_state: State,
+                          env: BehaviorEnv) -> Tuple[Array, bool]:
+        nonlocal plan_executed_forwards
+        nonlocal tried_closing_gripper
+        done_bit = False
+        if (not plan_executed_forwards and not tried_closing_gripper):
+            # Step thru the plan to execute Grasping
+            # phases 1 and 2
+            ret_action = get_delta_low_level_hand_action(
+                env,
```

Comment:
is there any hope of passing in some more specific things than env here? it's okay if not

Reply:
I think we can, but I'd like to test things before merging that in. I'll split this off into a new PR that we'll open once this is merged; I've opened an issue so we don't lose track of this in the meantime #273